common/src/main/java/org/openmrs/analytics/ParquetUtil.java

```diff
@@ -80,32 +140,77 @@ synchronized private AvroConverter getConverter(String resourceType) {
+		}
+		String bestFileName = String.format("output-streaming-%05d", bestIndex);
+		Path bestFilePath = new Path(Paths.get(getParquetPath(), resourceType).toString(), bestFileName);
+		log.info("Creating new Parguet file " + bestFilePath);
+		return bestFilePath;
+	}
+	
+	synchronized private void craeteWriter(String resourceType) throws IOException {
+		// TODO: Find a way to convince Hadoop file operations to use `fileSystem` (needed for testing).
+		AvroParquetWriter.Builder<GenericRecord> builder = AvroParquetWriter.builder(bestOutputFile(resourceType));
+		// TODO: Adjust other parquet file parameters for our needs or make them configurable.
+		if (rowGroupSize > 0) {
+			builder.withRowGroupSize(rowGroupSize);
+		}
+		ParquetWriter<GenericRecord> writer = builder.withSchema(getResourceSchema(resourceType)).build();
+		writerMap.put(resourceType, writer);
+	}
+	
 	/**
-	 * This is to fetch a writer for a given `resourceType`; internally it only once create a writer for
-	 * a `resourceType` and on subsequent calls, return the same object. The base path comes from
-	 * `file.parquetPath` system property. NOTE: This should be removed once we move the streaming
-	 * pipeline to Beam; the I/O should be left to Beam similar to the batch mode.
+	 * This is to write a FHIR resource to a Parquet file. This automatically handles file creation in a
+	 * directory named after the resource type (e.g., `Patient`) with names following the
+	 * "output-streaming-ddddd" pattern where `ddddd` is a string with five digits. NOTE: This should
+	 * not be used in its current form once we move the streaming pipeline to Beam; the I/O should be
+	 * left to Beam similar to the batch mode.
 	 */
-	synchronized public ParquetWriter<GenericRecord> getWriter(String resourceType) throws IOException {
+	synchronized public void write(Resource resource) throws IOException {
 		Preconditions.checkNotNull(getParquetPath());
+		Preconditions.checkNotNull(resource.fhirType());
+		String resourceType = resource.fhirType();
 		if (!writerMap.containsKey(resourceType)) {
-			AvroParquetWriter.Builder<GenericRecord> builder = AvroParquetWriter
-			        .builder(new Path(getParquetPath() + resourceType));
-			// TODO adjust parquet file parameters for our needs or make them configurable.
-			ParquetWriter<GenericRecord> writer = builder.withSchema(getResourceSchema(resourceType)).build();
-			writerMap.put(resourceType, writer);
+			craeteWriter(resourceType);
 		}
-		return writerMap.get(resourceType);
+		final ParquetWriter<GenericRecord> parquetWriter = writerMap.get(resourceType);
+		parquetWriter.write(this.convertToAvro(resource));
 	}
 	
-	synchronized public void closeAllWriters() {
+	synchronized private void flush(String resourceType) throws IOException {
+		ParquetWriter<GenericRecord> writer = writerMap.get(resourceType);
+		if (writer != null && writer.getDataSize() > 0) {
+			writer.close();
```

Comment:
are we also closing the writer when we experience the expectations above?

Reply:
Sorry which expectations do you mean? Here, we are closing any `writer` that is not null and non-empty.

Reply:
sorry meant to say IOException